Cleanup of old images, changing base image to Ubuntu.

ivannaranjo opened the pull request:
This PR changes the base image for all runtime images to use the `ubuntu:16.04` image.

We're also removing all of the old, and unsupported, images from the build process. Since that now simplifies our image set quite a bit we can now build all of the images with a single `cloudbuild.yaml`. This PR also does that, greatly simplifying our maintenance story moving forward.

The PR is also simplifying the `cloudbuild.yaml` file, there will not be a need to modify it every time there's a new .NET Core release. Only the files that compose the runtime image should be modified, namely, the `Dockerfile` to point to the updated runtime files and the test app, so we can verify with the latest bits.

ivannaranjo commented on runtimes/README.md:
```diff
@@ -1,22 +1,26 @@
-* The `versions` directory, which contains a directory for each supported minor version. Each directory contains the Dockerfile to build the image for the minor version.
-* The `structural_tests` directory, which again contains a directory for each supported minor version. Each directory contains a file called `aspnet.yaml` which defines what structural tests to run as part of the build for that minor version. These tests typically check that the right dotnet binary is included in the image and the licenses for all packages used in building the image.
-* The `functional_tests` directory, which contains the functional tests (unit tests) for each of the minor versions of the runtime. These tests are implemented as a simple console app that prints "Hello World!" to stdout. The build script will check that the app suceeded to run (had an exit code of 0).
+This direcotry contains the build definition for all of the supported .NET Core runtimes. The main structure is:
```
Done.

ivannaranjo commented on runtimes/README.md:
```diff
@@ -1,22 +1,26 @@
+This direcotry contains the build definition for all of the supported .NET Core runtimes. The main structure is:
+* The `cloudbuild.yaml` file contains the instructions to build and test all of the runtime images. This `cloudbuild.yaml` needs two substitutions to succeed, `_DOCKER_NAMESPACE` which is the name of the Docker repository where to store the image, and `_TAG` which is the tag to append to the image name to make it unique, typically this `_TAG` will be date based.
+* The `versions` directory contains the definition for each .NET Core version supported. In each version you will find:
+  + The `image` directory, which dontains the `Dockerfile` that defines the runtime image.
```
Done.

ivannaranjo commented on runtimes/README.md:
```diff
@@ -1,22 +1,26 @@
+* The `versions` directory contains the definition for each .NET Core version supported. In each version you will find:
+  + The `image` directory, which dontains the `Dockerfile` that defines the runtime image.
+  + The `structural_tests` directory, which contains the file called `aspnet.yaml`, this file defines what structural tests to run as part of the build for that version. These tests typically check that the right dotnet binary is included in the image and the licenses for all packages used in building the image.
+  + The `functional_tests` directory, which contains the functional tests (unit tests) for the runtime. These tests are implemented as a simple console app that prints "Hello World <version>!" to stdout. The build script will check that the app suceeded to run (had an exit code of 0) and that is has the expected output.
```
Done.

ivannaranjo commented on runtimes/README.md:
```diff
@@ -1,22 +1,26 @@
 
 ## Updating tests
-To update the tests, edit the source code in the `functional_tests/apps` directory and then run the [`udpate_runtime_functional_tests.sh`](../tools/update_runtime_functional_tests.sh) script on the corresponding `aspnetcore-<version>` directory. The script will take care of building the app and publishing it to the `functional_tests/published` directory for you.
+To update the tests, edit the source code in the `functional_tests/app` directory and then run the [`udpate_runtimes_tests.sh`](../tools/update_runtimes_tests.sh). The script will take care of building all of the apps and publishing it to the `functional_tests/published` directory for each supported runtime.
```
Done.

ivannaranjo commented on runtimes/cloudbuild.yaml:
```diff
@@ -0,0 +1,90 @@
+  waitFor: ['-']
+
+# .NET Core 1.0.x
+- name: gcr.io/cloud-builders/docker
```
Exactly. As we will only be using the latest version there's no need to refer o the `.x` part. That way you don't need to update the `cloudbuild.yaml` file when updating the runtimes.

ivannaranjo commented on runtimes/versions/aspnetcore-1.0/functional_tests/app/obj/Debug/netcoreapp1.0/app.AssemblyInfo.cs:
```diff
@@ -0,0 +1,17 @@
+//------------------------------------------------------------------------------
```
Oh no, I will cleanup the files. Will fix the `.gitignore` file for that.

ivannaranjo commented on runtimes/versions/aspnetcore-1.0/functional_tests/app/bin/Debug/netcoreapp1.0/app.runtimeconfig.dev.json:
```diff
@@ -0,0 +1,7 @@
+{
```
This file should be added. Will remove, see below.

ivannaranjo commented on runtimes/versions/aspnetcore-1.0/functional_tests/app/obj/app.csproj.nuget.cache:
```diff
@@ -0,0 +1,5 @@
+{
```
Nope.

ivannaranjo commented on runtimes/versions/aspnetcore-1.0/image/Dockerfile:
```diff
@@ -12,15 +12,15 @@
 # limitations under the License.
 
-FROM gcr.io/google_appengine/debian8
+FROM ubuntu:16.04
```
Good question. There is no GCR image for ubuntu, I presume for reasons beyond what we discuss here. Will contact the runtimes team to see if there's a better image, but doubt it.

ivannaranjo commented on tools/README.md:
```diff
@@ -2,9 +2,21 @@
+The [`build_runtimes.sh`](./build_runtimes.sh) script is a helper script to build all runtimes stored in the repo from a single command. It will also tag them if possible with the version numbers.
+
+## The build_builder.sh script.
+The [`build_builder.sh`](./build_builder.sh) script is a helper swcript to build the builder image. This script will also tag the builder as `latest` after it is done.
```
Done.

ivannaranjo commented:
PTAL

ILMTitan commented on tools/README.md:
```diff
@@ -2,9 +2,21 @@
 This script does not perform cleanup, the idea being that you might want to need to debug tests, re-run the tests manually after you are done, etc... To clean up you can just use `gcloud` to delete the versions created by this script. The versions are named after the directory name.
+
+## The build_all.sh script.
+The [`build_all.sh`](./build_all.sh) script is a helper script to build all runtime images and the builder image in a single command. It will use the currently selected GCP project by default, but you can override the Docker namespace to use in the parameter.
```
> It will use the currently selected GCP project by default, but you can override the Docker namespace to use in the parameter.

This sentence confused me. I think the phrase at the end should be "with the first parameter". I would also drop the clause about using the current GCP project, as it took me a bit to realize the connection between that and the Docker namespace. Instead, either repeat the last sentence in the `submit_build.sh` description, or refer to it.

ivannaranjo replied:
Reworked the sentence.

ILMTitan commented on tools/build_runtimes.sh:
```diff
@@ -0,0 +1,55 @@
+
+# If no repo is given get it from the ambient project.
+if [ -z "${1:-}" ]; then
+    readonly project_id=$(gcloud config list core/project --format="csv[no-heading](core)" | cut -f 2 -d '=')
```
Optional:

`gcloud config list core/project --format="csv[no-heading](core)" | cut -f 2 -d '='`
 ->
 `gcloud config list --format="value(core.project)"`

ivannaranjo replied:
Thank you for this, simplified the scripts with this.f

ivannaranjo commented on README.md:
````diff
@@ -1,30 +1,30 @@
 ```
 
-This will produce a directory, typically under `bin/release/netcoreapp1.0/publish/`. This is the directory where the `Dockerfile` for app's image should be placed.
+This will produce a directory, for .NET Core 2.0 it will be under `bin/release/netcoreapp2.0/publish/`. This is the directory where the `Dockerfile` for app's image should be placed.
````
Reworked the sentence a little bit, take a look.

ivannaranjo commented:
I have addressed the comments, fixed the READMEs and reworked the problematic sentences a bit. I have also refactored the scripts to share a bit more of code. These scripts are used for development of the images, usually not building them in production but it is good to clean them.

As for the versioning, as it turns out we really only care about the latest version on each branch of .NET Core. This means that eventually we will move towards a model in which we will use only the 2.0, 1.1 and 1.0 tags and avoid using the other tags anyway. In short, the actual patch version of the runtime that we are wrapping is not as important as we thought. It is still important to keep these in the `aspnetcore.yaml` file so we can detect whether we have a compatible runtime, but not for direct use.

ILMTitan commented:
LGTM
